### PR TITLE
Issue/1825 – Schema support for core REST endpoints (WIP)

### PR DESCRIPTION
**Pull request by DrewAPicture**
Issue: #1825

* [x] Affiliate schema
* [x] Creative schema
* [x] Payout schema
* [x] Referral schema
* [x] Visit schema

### CODE REVIEW

File: includes/REST/v1/class-affiliates-endpoints.php [hunk 1]
```diff
@@ -274,4 +274,102 @@ public function get_collection_params() {
...
+				'status'          => array(
+					'description' => __( 'The affiliate status.', 'affiliate-wp' ),
+					'type'        => 'string',
+				),
+				'earnings'        => array(
+					'description' => __( 'Affiliate earnings.', 'affiliate-wp' ),
+					'type'        => 'float',
+				),
+				'referrals'       => array(
+					'description' => __( 'The number of referrals associated with the affiliate.', 'affiliate-wp' ),
```
Comment (pippinsplugins) [inline includes/REST/v1/class-affiliates-endpoints.php]: Let's change this to `The number of paid referrals associated with the affiliate.`

File: includes/REST/v1/class-payouts-endpoints.php [hunk 2]
```diff
@@ -251,4 +251,62 @@ public function get_collection_params() {
...
+				'amount'        => array(
+					'description' => __( 'Total referrals amount for the payout.', 'affiliate-wp' ),
+					'type'        => 'float',
+				),
+				'payout_method' => array(
+					'description' => __( 'Method used to process the payout.', 'affiliate-wp' ),
+					'type'        => 'string',
+				),
+				'status'        => array(
+					'description' => __( 'The affiliate status.', 'affiliate-wp' ),
```
Comment (pippinsplugins) [inline includes/REST/v1/class-payouts-endpoints.php]: Should be `payout` not `affiliate`

File: includes/REST/v1/class-referrals-endpoints.php [hunk 3]
```diff
@@ -251,4 +251,79 @@ public function get_collection_params() {
...
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/schema#',
+			'title'      => $this->get_object_type(),
+			'type'       => 'object',
+			// Base properties for every affiliate.
```
Comment (pippinsplugins) [inline includes/REST/v1/class-referrals-endpoints.php]: `every affiliate` should be (I think) `every object`. Here and above.

File: includes/REST/v1/class-referrals-endpoints.php [hunk 4]
```diff
@@ -251,4 +251,79 @@ public function get_collection_params() {
...
+				'campaign'     => array(
+					'description' => __( 'Campaign associated with the referral.', 'affiliate-wp' ),
+					'type'        => 'string',
+				),
+				'reference'    => array(
+					'description' => __( 'Referral reference (usually a link to a specific sale).', 'affiliate-wp' ),
+					'type'        => 'string',
+				),
+				'products'     => array(
+					'description' => __( 'Products associated with the referral', 'affiliate-wp' ),
```
Comment (pippinsplugins) [inline includes/REST/v1/class-referrals-endpoints.php]: Missing a `.` at the end.

**Comment by TimothyBJacobs** [comment]
The `rate`, `rate_type`, and `payment_email` should each have their `type` set to `object` not `string`.

**Comment by DrewAPicture** [comment]
@pippinsplugins I've addressed all of your feedback.

@TimothyBJacobs Good catch!

**Pull request merged**